File: scrapers/linkedin.py
"""
LinkedIn Jobs guest API scraper (no authentication required).
Uses LinkedIn's public job search endpoint that returns HTML fragments.
"""
import logging
import time
import re
import requests
from bs4 import BeautifulSoup
from datetime import datetime, timedelta
from .base import Job
from config import DEFAULT_HEADERS, REQUEST_TIMEOUT, REQUEST_DELAY, SEARCH_KEYWORDS

logger = logging.getLogger(__name__)

# LinkedIn guest jobs search API
_SEARCH_URL = (
    "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search"
    "?keywords={keywords}&f_WT=2&f_JT=P%2CC%2CF%2CT&start={start}&count=25"
)
# f_WT=2 = remote
# f_JT = P(part-time), C(contract), F(full-time), T(temporary)


def scrape_linkedin(keywords: list[str] | None = None) -> list[Job]:
    if keywords is None:
        keywords = SEARCH_KEYWORDS

    results: list[Job] = []
    seen: set[str] = set()

    for kw in keywords[:5]:  # limit to top 5 keywords to avoid rate limiting
        jobs = _search_linkedin(kw)
        for job in jobs:
            if job.url in seen:
                continue
            seen.add(job.url)
            results.append(job)
        time.sleep(REQUEST_DELAY * 2)  # extra polite for LinkedIn

    logger.info(f"[LinkedIn] Found {len(results)} relevant jobs")
    return results


def _search_linkedin(keyword: str, max_pages: int = 2) -> list[Job]:
    results = []
    encoded = keyword.replace(" ", "%20")

    for page in range(max_pages):
        start = page * 25
        url = _SEARCH_URL.format(keywords=encoded, start=start)

        try:
            resp = requests.get(
                url,
                headers={**DEFAULT_HEADERS, "Accept": "text/html"},
                timeout=REQUEST_TIMEOUT,
            )
            if resp.status_code == 429:
                logger.warning("[LinkedIn] Rate limited, backing off")
                time.sleep(10)
                break
            if resp.status_code != 200:
                break
        except Exception as e:
            logger.warning(f"[LinkedIn] Request failed for '{keyword}': {e}")
            break

        jobs = _parse_linkedin_html(resp.text)
        if not jobs:
            break

        results.extend(jobs)
        time.sleep(REQUEST_DELAY)

    return results


def _parse_linkedin_html(html: str) -> list[Job]:
    soup = BeautifulSoup(html, "html.parser")
    jobs = []

    for card in soup.find_all("div", class_=re.compile(r"job-search-card|base-card")):
        title_el = card.find(["h3", "h4"], class_=re.compile(r"title|job-title"))
        company_el = card.find(class_=re.compile(r"company|subtitle"))
        location_el = card.find(class_=re.compile(r"location"))
        link_el = card.find("a", href=True)
        date_el = card.find("time")

        title = title_el.get_text(strip=True) if title_el else ""
        company = company_el.get_text(strip=True) if company_el else ""
        location = location_el.get_text(strip=True) if location_el else ""
        url = link_el["href"].split("?")[0] if link_el else ""
        if url and not url.startswith("http"):
            url = "https://www.linkedin.com" + url

        if not title or not url:
            continue

        # Parse relative date from datetime attribute
        posted = None
        if date_el and date_el.get("datetime"):
            try:
                posted = datetime.fromisoformat(date_el["datetime"])
            except Exception:
                pass

        # Determine job type from location text
        job_type = "remote" if "remote" in location.lower() else ""

        jobs.append(Job(
            title=title,
            company=company,
            url=url,
            source="LinkedIn",
            job_type=job_type,
            location=location,
            posted_date=posted,
        ))

    return jobs

File: scrapers/linkedin.js
import axios from "axios";
import { load } from "cheerio";
import { Job } from "./base.js";
import { DEFAULT_HEADERS, REQUEST_TIMEOUT_MS, REQUEST_DELAY_MS, SEARCH_KEYWORDS } from "../config.js";

// LinkedIn guest: Vietnam (geoId=104195383), remote+hybrid (f_WT=2,3), part-time+contract (f_JT=P,C)
const SEARCH_URL =
  "https://www.linkedin.com/jobs-guest/jobs/api/seeMoreJobPostings/search" +
  "?keywords={kw}&geoId=104195383&f_WT=2%2C3&f_JT=P%2CC&start={start}&count=25";

const DEFAULT_KEYWORDS = [
  "Power Platform", "Power BI", "SharePoint", "Power Automate",
  "Data Engineer", "Microsoft 365",
];

export async function scrapeLinkedIn(keywords = DEFAULT_KEYWORDS) {
  const results = [];
  const seen = new Set();
  const searchList = (keywords === SEARCH_KEYWORDS ? DEFAULT_KEYWORDS : keywords).slice(0, 6);

  for (const kw of searchList) {
    const jobs = await searchLinkedIn(kw);
    for (const job of jobs) {
      if (seen.has(job.url)) continue;
      seen.add(job.url);
      results.push(job);
    }
    await sleep(REQUEST_DELAY_MS * 2);
  }

  console.log(`[LinkedIn] ${results.length} relevant jobs`);
  return results;
}

async function searchLinkedIn(keyword, maxPages = 2) {
  const results = [];
  const encoded = encodeURIComponent(keyword);

  for (let page = 0; page < maxPages; page++) {
    const start = page * 25;
    const url = SEARCH_URL.replace("{kw}", encoded).replace("{start}", start);

    try {
      const res = await axios.get(url, {
        headers: { ...DEFAULT_HEADERS, Accept: "text/html" },
        timeout: REQUEST_TIMEOUT_MS,
      });
      if (res.status === 429) {
        console.warn("[LinkedIn] Rate limited");
        break;
      }
      const jobs = parseLinkedInHtml(res.data);
      if (!jobs.length) break;
      results.push(...jobs);
      await sleep(REQUEST_DELAY_MS);
    } catch (err) {
      console.warn(`[LinkedIn] '${keyword}' page ${page}: ${err.message}`);
      break;
    }
  }

  return results;
}

function parseLinkedInHtml(html) {
  const $ = load(html);
  const jobs = [];

  $("div.job-search-card, div.base-card").each((_, el) => {
    const card = $(el);
    const title = card.find(".base-search-card__title, h3, h4").first().text().trim();
    const company = card.find(".base-search-card__subtitle, .job-search-card__company-name").first().text().trim();
    const location = card.find(".job-search-card__location").first().text().trim();
    const linkEl = card.find("a[href]").first();
    let url = linkEl.attr("href") || "";
    if (url && !url.startsWith("http")) url = "https://www.linkedin.com" + url;
    url = url.split("?")[0];

    const dateEl = card.find("time");
    let postedDate = null;
    const datetime = dateEl.attr("datetime");
    if (datetime) {
      try { postedDate = new Date(datetime); } catch { /* ignore */ }
    }

    if (!title || !url) return;

    const jobType = location.toLowerCase().includes("remote") ? "remote" : "";

    jobs.push(
      new Job({
        title,
        company,
        url,
        source: "LinkedIn",
        jobType,
        location,
        postedDate,
      })
    );
  });

  return jobs;
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}
